### PR TITLE
fix(grouping): Make Stacktrace interface `is_url` check more restrictive

### PR DESCRIPTION
**Pull request by tkaemming**
This updates the `Stacktrace` interface's `is_url` method (and associated helper method) to return `True` for any URL that begins with a URL scheme as defined by [RFC 3986, Section 3.1](https://www.ietf.org/rfc/rfc3986.txt).

This does not account for:

1. protocol relative URLs (empty scheme),
2. URLs that do not have an "authority" component (incl. `mailto:`, `data:`, etc.),
3. other strings that are not syntactically valid URLs that coincidentally begin with a sequence of bytes that are in the form of a valid URL scheme,
4. bikeshed conversations about the distinction between a URI, a URL, and a URN.

### CODE REVIEW

**Comment by tkaemming** [comment]
```python
In [1]: from sentry.interfaces.exception import Exception

In [2]: Exception.to_python({'exc_omitted': None,
   ...:  'values': [{'mechanism': None,
   ...:              'module': None,
   ...:              'raw_stacktrace': None,
   ...:              'stacktrace': {'frames': [{'abs_path': 'applewebdata://00000000-0000-1000-8080-808080808080',
   ...:                                         'filename': 'applewebdata://00000000-0000-1000-8080-808080808080',
   ...:                                         'in_app': False}],
   ...:                             'frames_omitted': None,
   ...:                             'registers': None},
   ...:              'thread_id': None,
   ...:              'type': 'SyntaxError',
   ...:              'value': "Unexpected token '<'"}]}).get_hash()
Out[2]: ['SyntaxError', "Unexpected token '<'"]
```

File: tests/sentry/interfaces/test_stacktrace.py [hunk 1]
```diff
@@ -9,11 +9,19 @@
...
 from sentry.models import Event
 from sentry.testutils import TestCase
 
 
+def test_is_url():
+    assert is_url('http://example.org/') is True
+    assert is_url('https://example.org/') is True
+    assert is_url('file:///tmp/filename') is True
+    assert is_url('applewebdata://00000000-0000-1000-8080-808080808080') is True
+    assert is_url('data:,') is False  # no authority
```
Comment (tkaemming) [inline tests/sentry/interfaces/test_stacktrace.py]: I don't have a strong opinion about this last one but it's a lot simpler to parse if we're only concerned with URLs that contain an "authority" component (optional authentication, host, and optional port after a double slash.)

File: src/sentry/interfaces/stacktrace.py [hunk 2]
```diff
@@ -140,8 +140,11 @@ def is_newest_frame_first(event):
     return newest_first
 
 
+scheme_prefix_re = re.compile('^([a-z]+[a-z+-\.]*)://', re.IGNORECASE)
```
Comment (EvanPurkhiser) [inline src/sentry/interfaces/stacktrace.py]: Might be nice to see a comment here calling out that this is specifically tuned to RFC 3986 § 3.1

```
# Matches: https://tools.ietf.org/html/rfc3986#section-3.1
```

**Review by evanpurkhiser — APPROVED** [review]

File: src/sentry/interfaces/stacktrace.py (hunk 2 above)
Comment (tkaemming) [reply src/sentry/interfaces/stacktrace.py]: You just wanted to use the section glyph in a comment, be honest

Comment (EvanPurkhiser) [reply src/sentry/interfaces/stacktrace.py]: <kbd>◇6</kbd>

**Review by evanpurkhiser — APPROVED** [review]

**Review by mattrobenolt — APPROVED** [review]

**Comment by tkaemming** [comment]
Putting a little bit more thought and care into this before merging, because the result of JavaScript sourcemap expansion can result in a `webpack://` prefix being added to the `abs_path` attribute. This should be OK and not change hashing since I *think* that results in the `module` attribute also being added consistently (causing the `is_url` check to be skipped), but I need to verify that.

**Comment by mitsuhiko** [comment]
Also note that we use `app:///` for react native and many others.

**Comment by tkaemming** [comment]
Thanks for the heads up about `app://`. From what I can tell based on the test data, those don't have `module` attributes so this would break grouping for those frames.

I guess we'll have to switch back to a blacklist-based pattern and just keep adding to it as we find problematic cases like `applewebdata://`, unless we think we're able to put together a comprehensive whitelist. I'm not sure if any of you have opinions on that, @mitsuhiko, @mattrobenolt, et al.

**Comment by mitsuhiko** [comment]
I don't have huge opinions yet but I'm independently of this working on clarifying what our intended behavior is for paths and grouping. I think I have something to show on monday.

**Comment by tkaemming** [comment]
This was replaced by GH-6466.

**Pull request closed without merging**